pkg/services/ngalert/accesscontrol/silences.go

```diff
@@ -97,157 +97,219 @@ func NewSilenceService(ac ac.AccessControl, store RuleUIDToNamespaceStore) *Sile
+				namespacesByAccess[silWithFolder.folderUID] = hasAccess
 			}
-			namespacesByAccess[ns] = hasAccess
 		}
 		if hasAccess {
-			result = append(result, silence...)
+			result = append(result, silWithFolder.Silence)
 		}
 	}
 	return result, nil
 }
 
-// AuthorizeReadSilence checks if user has access to read a silence
+// AuthorizeReadSilence checks if user has access to read a silence.
 func (s SilenceService) AuthorizeReadSilence(ctx context.Context, user identity.Requester, silence *models.Silence) error {
-	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	canAll, err := s.authorizeReadSilencePreConditions(ctx, user)
 	if canAll || err != nil { // return early if user can either read all silences or there is error
 		return err
 	}
 
+	silWithFolder, err := s.withFolders(ctx, user.GetOrgID(), silence)
+	if err != nil || len(silWithFolder) != 1 {
+		return fmt.Errorf("resolve rule UID to folder UID: %w", err)
+	}
+
+	return s.authorizeReadSilence(ctx, user, silWithFolder[0])
+}
+
+// authorizeReadSilencePreConditions checks necessary preconditions for reading silences. Returns true if user can
+// read all silences. Returns error if user does not have access to read any silences.
+func (s SilenceService) authorizeReadSilencePreConditions(ctx context.Context, user identity.Requester) (bool, error) {
+	canAll, err := s.HasAccess(ctx, user, readAllSilencesEvaluator)
+	if canAll || err != nil { // return early if user can either read all silences or there is error
+		return canAll, err
+	}
+
 	can, err := s.HasAccess(ctx, user, readSomeSilenceEvaluator)
 	if err != nil {
-		return err
+		return false, err
 	}
 	if !can { // User does not have silence permissions at all.
-		return NewAuthorizationErrorWithPermissions("read any silences", readSomeSilenceEvaluator)
-	}
-	ruleUID := silence.GetRuleUID()
-	if ruleUID == nil {
-		return nil // no rule UID means that this is a general silence and at this point the user can read them
+		return false, NewAuthorizationErrorWithPermissions("read any silences", readSomeSilenceEvaluator)
 	}
+	return false, nil
+}
 
-	// otherwise resolve rule key to the action's scope
-	folderUID, err := s.ruleUIDToFolderUID(ctx, user.GetOrgID(), *ruleUID)
-	if err != nil {
-		return fmt.Errorf("resolve rule UID to folder UID: %w", err)
+// authorizeReadSilence checks if user has access to read a silence given precondition checks have passed.
+func (s SilenceService) authorizeReadSilence(ctx context.Context, user identity.Requester, silence *silenceWithFolder) error {
+	if silence.ruleUID == nil {
+		return nil // No rule metadata means that this is a general silence and at this point the user can read them
 	}
-	if folderUID == "" { // if we did not find folder by rule UID then it does not exist.
-		return NewAuthorizationErrorGeneric(fmt.Sprintf("read silence for rule %s", *ruleUID))
+
+	if silence.folderUID == "" { // if we did not find folder by rule UID then it does not exist.
```

Comment:
If I read this correctly, in a situation when a rule is assigned to a folder that does not exist (which can happen), nobody will be able to see this silence, even "admin". I think we should return it if user can read all silences (ideally, when wildcard folder scope too)